### ISSUE TITLE
macOS+Lua notification center events not working 

### ISSUE THREAD

**Issue by wahern** [yue/yue#135]
Yue Version: 0.10.2
Operating System: macOS

Notification center event handlers seem to never be called. I've tried assigning to gui.notificationcenter.onnotificationshow, gui.notificationcenter.onnotificationclose, gui.notificationcenter.onnotificationclick, and gui.notificationcenter.onnotificationaction. Calling the :connect method on those slots (e.g. .onnotificationshow:connect()) fails with "attempt to index a nil value". Reading through the C++ code it looks like something should be there (all the names seem correct), but I don't understand the code well enough to guess what the issue might be.

**Comment by wahern** [comment]
I think I figured it out: NSUserNotificationCenter was replaced by UNUserNotificationCenter in macOS 10.14, and NSUserNotificationCenter stopped working in macOS 11.0.

If I have time (and I don't at the moment) I'll see if I can put together a patch to make the switch.

**Comment by wahern** [comment]
Regarding gui.lifetime.onactivate and gui.lifetime.onready issues I mentioned in the July 13th comment: I managed to get these working, but only by explicitly invoking `+[NSApplication sharedApplication]` (through a tiny FFI wrapper based on objc_msgSend) before loading Yue (i.e. luaopen_yue_gui). This code works,

```
  require"lucky.apple.objc".msgsend("NSApplication", "sharedApplication")
  local yue = require"yue.gui"
  yue.lifetime.onactivate:connect(function ()
    logger.notice"received lifetime onactivate event"
  end)
  yue.lifetime.onready:connect(function ()
    logger.notice"received lifetime onready event"
  end)
```

but the following code does not work (no events received, not even by the Objective-C handlers),

```
  local yue = require"yue.gui"
  require"lucky.apple.objc".msgsend("NSApplication", "sharedApplication")
  yue.lifetime.onactivate:connect(function ()
    logger.notice"received lifetime onactivate event"
  end)
  yue.lifetime.onready:connect(function ()
    logger.notice"received lifetime onready event"
  end)
```

I have no idea why this works or how it works; other Yue initialization code already invokes that, through `NSApp`. I did add a bunch of printf statements to track various code paths and state changes in Yue, including whether and how the NSApplication object address changed based on ordering of any explicit sharedApplication call, but in starting to write this comment I realize that I've forgotten what I ended up seeing.

The yue.gui.Lifetime documentation says `When using this class, you must create Lifetime before creating State`, and that's what luaopen_yue_gui seems to do:

```
extern "C" int luaopen_yue_gui(lua::State* state) {
  // The exports table.
  lua::NewTable(state);

  // Initialize nativeui.
  lua::Push(state, "__lifetime_ptr");
  lua::NewUserData<nu::Lifetime>(state);
  lua_rawset(state, -3);
  lua::Push(state, "__state_ptr");
  lua::NewUserData<nu::State>(state);
  lua_rawset(state, -3);
```

But clearly there's something I'm missing. Note that I'm statically linking Yue, so the Yue module is already statically initialized in the process before calling require. The only difference in my two code blocks is whether `+[NSApplication sharedApplication]` is invoked before or after luaopen_yui_gui.

Leaving this comment in case other people run into issues with yue.gui.Lifetime events, but happy to see this ticket closed as it concerns the now-fixed notification center bindings.
